src/processors/default/ipfsMediaUploader.ts

```diff
@@ -0,0 +1,77 @@
+
+import { urlSource } from 'ipfs-http-client';
+import _ from 'lodash';
+
+
+import { Table } from '../../db/db';
+import { artworkNotOnIpfs, audioNotOnIpfs } from '../../triggers/ipfs';
+import { IPFSFile } from '../../types/ipfsFIle';
+import { Clients, Processor } from '../../types/processor';
+import { ProcessedTrack } from '../../types/track';
+import { rollPromises } from '../../utils/rollingPromises';
+
+function processorFunction(sourceField: 'lossyAudioURL' | 'lossyArtworkURL', replaceField: 'lossyAudioIPFSHash' | 'lossyArtworkIPFSHash') {
+  return async (tracks: ProcessedTrack[], clients: Clients) => {
+    const updates: ProcessedTrack[] = [];
+    const ipfsFiles: IPFSFile[] = [];
+
+    const existingFiles = await clients.db.getRecords<IPFSFile>(Table.ipfsFiles, [
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      ['whereIn', ['url', tracks.map(track => track[sourceField])]]
+    ]);
+    const fileByUrl = _.keyBy(existingFiles, 'url')
+    
+    const processTrack = async (track: ProcessedTrack) => {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      const url = track[sourceField];
+      try {
+        const source = urlSource(url)
+       
+        const fileForUrl = fileByUrl[url];
+
+        if (!fileForUrl){
+          const file = await clients.ipfs.client.add(source, {
+            pin: false
+          });
+          const cid = file.cid.toString();
+          updates.push({
+            ...track,
+            [replaceField]: cid
+          })
+          ipfsFiles.push({
+            url,
+            cid
+          })
+        } else {
+          updates.push({
+            ...track,
+            [replaceField]: fileForUrl.cid
+          })
+        }
+      } catch (e: any){
+        ipfsFiles.push({ url, error: e.message });
+      }
+    }
+    
+    await rollPromises<ProcessedTrack, void, void>(tracks, processTrack, 300, 60)
```

Comment:
how do you imagine/expect things like large files, timeouts, etc to get handled in production and operationally?

Reply:
There is a timeout variable in the config, if the request timeout is reached then an error will be written to the error column of the ipfsFiles table and that track will be skipped until that error is cleared

Reply:
hmm didn't see which timeout variable applies to this?

also, do you know if it handles each of the cases correctly, eg:
 - (A) add a track from a source that is really slow, so it takes very long to add but it is stable and makes progress and so should correctly finish
 - (B) add a track from a source that is fast, but the track is a 40min audiovisual high quality video, so like 1gb, so it takes long but is fast and stable and so should correctly finish
 - (C) add a track from a source that is unavailable or cuts out, so is a genuine error and should timeout and error

I suppose we should have a maximum of a few minutes irrespective of the above scenarios, since we don't want the processor to block on one big track, so even if it's case A or B maybe it should still timeout and we should handle via error?

Reply:
so far I've only tested with case B and C. I imagine A would cause a timeout too but I will try find a suitable example to test that with. Let me know if you are aware of any

Reply:
seems fine for now. if we encounter it in prod we can deal with it then